### ISSUE TITLE
Using require.ensure, TypeScript dependencies are not included in specified chunk

### ISSUE THREAD

**Issue by scameron** [webpack/webpack#14720]
# Bug report

**What is the current behavior?**

When using `require.ensure` along with a specified chunkName to require TypeScript code, the transpiled code does not get added to the specified chunk, but instead to the chunk of the caller.  This massively breaks lazy loading behaviour in our code base.

**If the current behavior is a bug, please provide the steps to reproduce.**

Please see the Github repo here for a repro case build with both Webpack 3 (the expected behaviour) and Webpack 5 (the unexpected behaviour): https://github.com/scameron/webpack-issue.

**What is the expected behavior?**

All code for the `require.ensure` should be added to the specified chunk, regardless of its origin.

**Comment by alexander-akait** [comment]
It is breaking change, please read changelog

**Comment by scameron** [comment]
> It is breaking change, please read changelog

I've looked in all of the migration docs and changelogs I can find but maybe I'm missing it.  Which log are you referring to?

**Comment by alexander-akait** [comment]
Try this https://webpack.js.org/configuration/module/#moduleparserjavascript

**Comment by alexander-akait** [comment]
No solutions, please rewrite your code on ESM

**Comment by scameron** [comment]
`require.ensure` works fine with JS dependencies.  The issue seems to only happen with TS (or, at least, with non-JS).

`require.ensure` worked with TS in previous releases and that call is not (yet) deprecated so why isn't this a valid bug to fix?

**Comment by alexander-akait** [comment]
`require.ensure` is frozen due it is non standard, but give me time I will look deeply

**Comment by alexander-akait** [comment]
It happens not only with TS, using JS webpack still move `ModuleB` in `chunk1`

**Comment by scameron** [comment]
> It happens not only with TS, using JS webpack still move `ModuleB` in `chunk1`

What do you mean by "using JS" ?   `ModuleC` is JS and it gets moved to `chunk2`.   What's the difference with `ModuleB` if not the fact that it is TS?

**Comment by alexander-akait** [comment]
If you rename `ModuleB.ts` to `ModuleB.js` problem persists

**Comment by alexander-akait** [comment]
Oh, if you change:
```
const ModuleB = require("./ModuleB").ModuleB;
ModuleB.testFunction();
```

on:

```
const ModuleB = require("./ModuleB");
ModuleB.testFunction();
```

All works fine

**Comment by scameron** [comment]
That actually makes more sense that it is related to the export and not to the language of the originating code.  Good one!

Honestly, though, we have an enormous code base and while we have it in our backlog to move to ESM, anything requiring changing our requires will be an enormous effort.

Since this worked in a previous version, would it be possible to investigate this as a bug fix, even though there are workaround?

The fact that this doesn't fail explicitly, but rather just silently changes the internal loading behaviour, is pretty insidious.  I bet there are a lot of legacy projects around that are affected by this.  You wouldn't believe what this issue did to our runtime code footprint.  We are seeing a huge amount of code eager-loaded where it used to be lazy-loaded on demand.

**Comment by alexander-akait** [comment]
Yes, there is a problem somewhere, unfortunately, this is not at all a priority for us, and I'm afraid that the fix will not be in the near future, anyway you can investigate it and send a fix, most likely the problem is not so difficult

**Comment by scameron** [comment]
I think I see the problem. In `CommonJsImportsParserPlugin.js` there are a couple of places where `addDependency` is called on the module instead of the current parser state. Looks like an easy fix so I’ll try to do a PR for that with a test.

**Comment by sokra** [comment]
fix is here: #14725

**Comment by sokra** [comment]
Good find @scameron
